specs/plugin-extensibility.md

```diff
@@ -0,0 +1,277 @@
+
+## Terminology
+
+* **Plugin Publisher** - A user, organization, open source project or 3rd party vendor that creates a Notation plugin for internal or public distribution.
+* **Plugin** - A component external to Notation that can integrate as one of the steps in  Notation’s workflow for signature generation or verification.
+* **Default provider** - Signing and verification mechanisms built into Notation itself to provide default experience without requiring to install/configure additional plugins. ***[We are yet to define what is included in the default experience]***.
+
+## Plugin mechanism
+
+### Requirements
+
+* A plugin publisher MUST be able to distribute and patch a Notation plugin independently of Notation’s release cycle.
+* The plugin mechanism MUST work across commonly used OS platforms like Linux, Windows and macOS.
+* The plugin interface contract MUST be versioned. This version is different from Notation's release version. It will be used to accomodate for additional capabilities and contract changes post initial release of Notation.
+* A plugin MAY implement a subset of capabilities (features) available in plugin contract. E.g A plugin may implement signing feature, but not verification.
+* Notation and plugins MAY be updated independently in an environment.
+  * Notation MUST work with a plugin that implements a matching or lower minor version of the plugin contract. Notation SHALL NOT support using a plugin with higher version of plugin contract.
+  * A plugin MUST support a single plugin contract version, per major version.
+
+Notation will invoke plugins as executable, pass parameters using command line arguments, and use standard IO streams to pass request/response payloads. This mechanism is used as Go language (used to develop [Notation library](https://github.com/notaryproject/notation-go-lib)) does not have a [in built support](https://github.com/golang/go/issues/19282) to load and execute plugins that works across OS platforms. Other mechanisms like gRPC require every plugin to be implemented as a service/deamon.
+
+### Plugin installation and config
+
+* Plugin publisher will provide instructions to download and install the plugin. Plugins intended for public distribution should also include instructions for users to verify the authenticity of the plugin.
+* Each plugin executable and dependencies are installed under directory `~/.notation/plugins/{plugin-name}` with an executable under that directory `~/.notation/plugins/{plugin-name}/notation-{plugin-name}`. The executable can be a shim which calls plugin dependecies installed elsewhere on the file system.
+* To use a plugin for signing, the user associates the plugin as part of registering a signing key. E.g.
+  * `notation key add --name "mysigningkey" --id "keyid" --plugin "com.amazonaws.signer.nv2plugin"`
```

Comment:
If there are multiple plugins in the plugins-dir, does `com.amazonaws.signer.nv2plugin` map to a particular plugin?